### ISSUE TITLE
Multithreading issues

### ISSUE THREAD

**Issue by hameerabbasi** [Quansight-Labs/uarray#219]
Consider the following code:

```python
with ua.set_backend(numpy_backend), ua.set_backend(dask_backend):
    x = np.arange(0, 5000, 1)
    y = np.cos(x)
    y.compute()
```

This faces the following error (which I'm fairly sure is multithreading-related):
<details>
  <summary>Traceback</summary>
  <p>

```
Traceback (most recent call last):
  File "/Users/hameerabbasi/Quansight/udiff/sandbox.py", line 15, in <module>
    y.arr.compute()
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/base.py", line 165, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/base.py", line 436, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/threaded.py", line 81, in get
    **kwargs
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/local.py", line 486, in get_async
    raise_exception(exc, tb)
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/local.py", line 316, in reraise
    raise exc
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/local.py", line 222, in execute_task
    result = _execute_task(task, data)
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/core.py", line 119, in _execute_task
    return func(*args2)
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/optimization.py", line 982, in __call__
    return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/core.py", line 149, in get
    result = _execute_task(task, cache)
  File "/Users/hameerabbasi/miniconda3/envs/uarray/lib/python3.7/site-packages/dask/core.py", line 119, in _execute_task
    return func(*args2)
uarray.BackendNotImplementedError: No selected backends had an implementation for this function.
```
</p>
</details>

If we set a breakpoint, we see that new threads have been spawned.

**Comment by peterbell10** [comment]
So you are setting the thread's local backend and then `unumpy`'s [dask backend](https://github.com/Quansight-Labs/unumpy/blob/208146ed930d62dbacb529f936c5ddcf10cc0255/unumpy/dask_backend.py#L19-L25) calls `uarray` functions from a new thread where there is no backend set.
```python
da.map_blocks(func, *args, **kwargs)
```
where `func` is `unumpy.ufunc.__call__`.

**Comment by hameerabbasi** [comment]
Aren't thread-local variables copied over to new threads? Is there a way to copy them over?

**Comment by peterbell10** [comment]
No, they are default-initialised on the child thread. I'm not aware of any way to automatically copy them over when spawning a thread.

**Comment by hameerabbasi** [comment]
Ah, the Python implementation I had used context variables which do have that behavior.

**Comment by hameerabbasi** [comment]
Okay, so I found this: https://stackoverflow.com/questions/23555612/can-a-c11-thread-local-variable-inherit-its-initial-value-from-the-parent-thre

**Comment by peterbell10** [comment]
In that case they are just initializing the `thread_local` from a global static variable. Firstly, that only works for one "master" thread and not arbitrarily nested thread launches. Secondly, the thread's values aren't copies over at thread launch but at first use. So, the threads will race if the parent modifies the global state after launching a thread.

**Comment by hameerabbasi** [comment]
> Secondly, the thread's values aren't copies over at thread launch but at first use. So, the threads will race if the parent modifies the global state after launching a thread.

I'm not super concerned about this scenario.

> Firstly, that only works for one "master" thread and not arbitrarily nested thread launches.

Are you sure this is the case?

**Comment by peterbell10** [comment]
>I'm not super concerned about this scenario.

It means that the backends available in the child is non-deterministic. That sounds like a huge issue to me.

>Are you sure this is the case?

There's only one instance of the global state so how could you possibly handle nested thread launches? You could generalise the code further by making all thread's contexts global and essentially manage the `thread_local` aspect manually. However, you still have no way to identify a parent-child relationship between the different threads.

**Comment by hameerabbasi** [comment]
> It means that the backends available in the child is non-deterministic. That sounds like a huge issue to me.

It's only an issue in this case:

```
spawn_child
set_backend
```
Which seems like an edge-case in this scenario to me. If we do the following, which is the usual usage pattern:

```
set_backend
    spawn_child
    join_child_thread
```

I see no problem.

> There's only one instance of the global state so how could you possibly handle nested thread launches? You could generalize the code further by making all thread's contexts global and essentially manage the thread_local aspect manually. However, you still have no way to identify a parent-child relationship between the different threads.

While its true that all threads are created equal, IIUC the `thread_local` variable already does this  for you (copying over the parent thread's `thread_local` state between thread launch and first use, I feel that's what the SO link says). I'd hate to be proven wrong on this point, because then it means there really is no way to fix this.

**Comment by hameerabbasi** [comment]
I suspect Dask is using `multiprocessing` instead of `threading`, which spawns a new process entirely, causing the issue.

**Comment by peterbell10** [comment]
>I see no problem.

Fork-join parallelism is not the only way to use threads. Even your dask backend doesn't fit into this pattern since the thread launch is deferred until you call `compute()`.

>While its true that all threads are created equal, IIUC the thread_local variable already does this for you (copying over the parent thread's thread_local state between thread launch and first use, I feel that's what the SO link says).

The SO link says nothing like that. They are using an initializing expression from a global variable:
```cpp
trace_level log_level = trace_level::log;
thread_local trace_level min_level = log_level;
```
So, `trace_level` is initialised with the current value of `log_level`.

**Comment by hameerabbasi** [comment]
> The SO link says nothing like that. They are using an initializing expression from a global variable

How about the following:

```
trace_level get_current_min_level() {
    return min_level;
}

thread_local trace_level min_level = get_current_min_level();
```

> Fork-join parallelism is not the only way to use threads. Even your dask backend doesn't fit into this pattern since the thread launch is deferred until you call `compute()`.

Well, `compute()` does join the children threads, so it's really like the following:

```
set_backend
    do_something
    spawn_child // inside .compute()
    join_child_thread // inside .compute()
```

**Comment by hameerabbasi** [comment]
Worst comes to worst, we can use this: https://docs.python.org/3/c-api/contextvars.html

**Comment by peterbell10** [comment]
>How about the following:

What about it? `get_current_min_level` would still need to know the value in the parent thread.

>Well, `compute()` does join the children threads

This pattern from the dask backend doesn't work
```python
with skip_backend(sys.modules[__name__]):
    return da.map_blocks(func, *args, **kwargs)
```

The `skip_backend` context has already completed by the time `compute()` is called.

**Comment by hameerabbasi** [comment]
Resolved by #220. We decided to go with closures that set the state.
